### BREAST CANCER DATASET

##### Introduction

Breast cancer diagnosis is a critical medical task, as misclassification especially false negatives which can lead to severe clinical consequences. This project aims to develop a robust machine learning pipeline to classify tumors as Benign (B) or Malignant (M) using the Breast Cancer dataset.

The primary objectives include:

- Cleaning and preprocessing the dataset

- Handling feature scaling and class imbalance

- Building and tuning Logistic Regression and Random Forest models

- Evaluating model performance using medically relevant metrics

- Selecting the best model based on recall and F1-score


#### Phase One: Data collection and preparation

In [2]:
# Import all ncessary libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [3]:
# Data ingestion: Load the dataset
breast_cancer_data = pd.read_csv('Breast_cancer_dataset.csv')
print("First five rows of the dataset\n")
breast_cancer_data.head(5)

First five rows of the dataset



id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

#### Inspect the dataset for missing values, duplicates, datatype and shape

In [4]:
# Check the data info
breast_cancer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [5]:
# Check the shape of the dataset
print(f"No of Rows: {breast_cancer_data.shape[0]}")
print(f"No of Columns: {breast_cancer_data.shape[1]}")

No of Rows: 569
No of Columns: 33


In [6]:
# Check for duplicate values
print(f"Duplicated values: {breast_cancer_data.duplicated().sum()}")

Duplicated values: 0


In [7]:
# Check for missing values
breast_cancer_data.isna().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed:

In [8]:
# Print the columns in the dataset
breast_cancer_data.columns

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32'],
      dtype='object')

#### Data Overview 
- The dataset uses 569 breast cancer prediction dataset
- Each prediction has 30 measurements that describes the medical properties of the features, and a diagnosis (target variable) that is either 'B -Benign' or 'M- Malignant'
- The goal is to understand what makes a lump in the breast cancerous or non-cancerous and prepare the data for Ml model that predicts the state of the lump.


#### Data cleaning: handle missing values

In [9]:
# Create a copy of the dataset
breast_cancer_df = breast_cancer_data.copy()

In [10]:
# Set the case id as the index
breast_cancer_df.set_index("id", inplace=True)

In [11]:
# Display the first two rows
breast_cancer_df.head(2)

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                       
842302         M        17.99         10.38           122.8     1001.0   
842517         M        20.57         17.77           132.9     1326.0   

        smoothness_mean  compactness_mean  concavity_mean  \
id                                                          
842302          0.11840           0.27760          0.3001   
842517          0.08474           0.07864          0.0869   

        concave points_mean  symmetry_mean  ...  texture_worst  \
id                                          ...                  
842302              0.14710         0.2419  ...          17.33   
842517              0.07017         0.1812  ...          23.41   

        perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                         
842302            184.6      2019.0            0.1622             0.6656   
842517            158.8      1956.0            0.1238             0.1866   

        concavity_worst  concave points_worst  symmetry_worst  \
id                                                              
842302           0.7119                0.2654          0.4601   
842517           0.2416                0.1860          0.2750   

        fractal_dimension_worst  Unnamed: 32  
id                                            
842302                  0.11890          NaN  
842517                  0.08902          NaN  

[2 rows x 32 columns]

In [12]:
# Handle the column with missing values (NAN)
breast_cancer_df = breast_cancer_df.drop('Unnamed: 32', axis=1)


In [13]:
# Confirm missing colunmn has been dropped
breast_cancer_df.isna().sum()

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

In [14]:
# Display the columns
breast_cancer_df.head(2)

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                       
842302         M        17.99         10.38           122.8     1001.0   
842517         M        20.57         17.77           132.9     1326.0   

        smoothness_mean  compactness_mean  concavity_mean  \
id                                                          
842302          0.11840           0.27760          0.3001   
842517          0.08474           0.07864          0.0869   

        concave points_mean  symmetry_mean  ...  radius_worst  texture_worst  \
id                                          ...                                
842302              0.14710         0.2419  ...         25.38          17.33   
842517              0.07017         0.1812  ...         24.99          23.41   

        perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                         
842302            184.6      2019.0            0.1622             0.6656   
842517            158.8      1956.0            0.1238             0.1866   

        concavity_worst  concave points_worst  symmetry_worst  \
id                                                              
842302           0.7119                0.2654          0.4601   
842517           0.2416                0.1860          0.2750   

        fractal_dimension_worst  
id                               
842302                  0.11890  
842517                  0.08902  

[2 rows x 31 columns]

In [15]:
# Check the values in the target column (diagnosis)
breast_cancer_df['diagnosis'].unique()

array(['M', 'B'], dtype=object)

#### Getting the data ready for EDA
- No duplicated values found.
- A column with invalid values was found and were handled (dropped).


In [16]:
# Convert the diagnosis 
breast_cancer_df = breast_cancer_df.replace({"diagnosis":{'M':'Malignant', 'B': 'Benign'}})

#### Save the cleaned data to CSV

In [17]:
breast_cancer_df.to_csv('cleaned_breast_cancer_dataset.csv', index=True)